File: AlbumLibrary/src/main/java/com/album/ui/widget/LoadMoreRecyclerView.java
package com.album.ui.widget;

import android.content.Context;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.AttributeSet;

/**
 * by y on 22/08/2017.
 */

public class LoadMoreRecyclerView extends RecyclerView {


    private int lastVisibleItemPosition;
    private LoadMoreListener loadingListener;

    public LoadMoreRecyclerView(Context context) {
        super(context);
    }


    public LoadMoreRecyclerView(Context context, AttributeSet attrs) {
        super(context, attrs);
    }


    public LoadMoreRecyclerView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
    }

    public void setLoadingListener(LoadMoreListener loadingListener) {
        this.loadingListener = loadingListener;
    }

    @Override
    public void onScrolled(int dx, int dy) {
        super.onScrolled(dx, dy);
        GridLayoutManager layoutManager = (GridLayoutManager) getLayoutManager();
        lastVisibleItemPosition = layoutManager.findLastVisibleItemPosition();
    }

    @Override
    public void onScrollStateChanged(int state) {
        super.onScrollStateChanged(state);
        LayoutManager layoutManager = getLayoutManager();
        int visibleItemCount = layoutManager.getChildCount();
        int totalItemCount = layoutManager.getItemCount();
        if (loadingListener != null && visibleItemCount > 0 &&
                state == RecyclerView.SCROLL_STATE_IDLE && lastVisibleItemPosition == totalItemCount - 1) {
            loadingListener.onLoadMore();
        }
    }


    public interface LoadMoreListener {
        void onLoadMore();
    }
}

File: AlbumLibrary/src/main/java/com/album/ui/widget/LoadMoreRecyclerView.kt
package com.album.ui.widget

import android.content.Context
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.RecyclerView
import android.util.AttributeSet

/**
 * by y on 22/08/2017.
 */

class LoadMoreRecyclerView : RecyclerView {


    private var lastVisibleItemPosition: Int = 0
    private var loadingListener: LoadMoreListener? = null

    constructor(context: Context) : super(context) 


    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) 


    constructor(context: Context, attrs: AttributeSet, defStyle: Int) : super(context, attrs, defStyle) 

    fun setLoadingListener(loadingListener: LoadMoreListener) {
        this.loadingListener = loadingListener
    }

    override fun onScrolled(dx: Int, dy: Int) {
        super.onScrolled(dx, dy)
        val layoutManager = layoutManager as GridLayoutManager
        lastVisibleItemPosition = layoutManager.findLastVisibleItemPosition()
    }

    override fun onScrollStateChanged(state: Int) {
        super.onScrollStateChanged(state)
        val layoutManager = layoutManager
        val visibleItemCount = layoutManager.childCount
        val totalItemCount = layoutManager.itemCount
        if (loadingListener != null && visibleItemCount > 0 &&
                state == RecyclerView.SCROLL_STATE_IDLE && lastVisibleItemPosition == totalItemCount - 1) {
            loadingListener!!.onLoadMore()
        }
    }


    interface LoadMoreListener {
        fun onLoadMore()
    }
}
